### PR TITLE
Fix ite type checking bug (#571)

### PR DESCRIPTION
**Pull request by leonardt**
Before, ite was allowing SInt and UInt to be used together, which should
not be allowed.  This fixes it by first updating the ite logic to use
the mux operator (which has improved type checking logic) and updating
the wireable logic to prevent UInt and SInt from being wireable (they
match structurally but we should error when trying to wire them
together).  I think we should eventually update the wireable logic to
allow the types to define their wireability themselves (rather than
housing the logic internal to magma since this won't extend to user
defined types).  However, since UInt/SInt are built-in to magma, this
should be fine for now.

### CODE REVIEW

File: tests/test_primitives/test_ite.py [hunk 1]
```diff
@@ -0,0 +1,22 @@
+import pytest
+import magma as m
+
+
+def test_adc_signed_unsigned():
+    Data = m.Bits[16]
+    UData = m.UInt[16]
+    SData = m.SInt[16]
+
+    with pytest.raises(TypeError) as e:
```
Comment (rsetaluri) [inline tests/test_primitives/test_ite.py:10]: Can we distill the test(s) for what we want to check? In particular it might be easier to read and maintain if the test just does the minimal thing, i.e. not wrapped in `combinational`. Perhaps one test to check (a) basic sint-uint wiring (see comment above), and one test for (b) mux'ing those.

File: magma/wireable.py [hunk 2]
```diff
@@ -8,6 +9,9 @@ def wireable(T1, T2):
     Returns true if T1 can be wired to T2
     """
     T1, T2 = magma_type(T1), magma_type(T2)
+    if ((issubclass(T1, UInt) and issubclass(T2, SInt)) or
```
Comment (rsetaluri) [inline magma/wireable.py:12]: what's the difference between `wireable(x, y)` and `x.is_wireable(y)`? It seems there's some fragmentation of this logic. Ok if it's the case, but ideally should document when to use one vs the other (or update names appropriately).

In this case, I tried wiring a uint to an sint and it worked so that seems to be inconsistent with the logic here? Is this only for muxing?

**Comment by cdonovick** [comment]
> think we should eventually update the wireable logic to allow the types to define their wireability themselves

I think this a good idea.  Add a `_wire_` method to the magma interfce.

And make the wire function do something basically like:
```Python
def wire(lhs, rhs):
    # ... Logic around the magma protocol ...
    try:
        res = lhs._wire_(rhs)
    except AttributeError:
        res = NotImplemented # or some magma specific singleton
    if res is NotImplemented:
        try:
            res = rhs._wire_(lhs)
        except AttributeError:
            res = NotImplemented # or some magma specific singleton
    if res is NotImplemented:
        raise WiringError()
```

File: magma/wireable.py (hunk 2 above)
Comment (leonardt) [reply magma/wireable.py:12]: They should be consistent, I can change it to call is_wireable on the types (related to Caleb's feedback)

File: tests/test_primitives/test_ite.py (hunk 1 above)
Comment (leonardt) [reply tests/test_primitives/test_ite.py:10]: Sure, the sint-uint wiring is the underlying problem, so we need to fix that.  then there's the fact that ite (used by combinational/sequential) isn't using the mux operator which defines the type checking rules for inference (so we can change it from using combinational to just calling ite directly)

File: magma/primitives/mux.py [hunk 3]
```diff
@@ -108,7 +107,7 @@ def _infer_mux_type(args):
             if issubclass(T, next_T):
                 # upcast
                 T = next_T
-            elif not wireable(next_T, T):
```
Comment (rsetaluri) [inline magma/primitives/mux.py:111]: can we get rid of `magma/wireable.py` then?

File: magma/bits.py [hunk 4]
```diff
@@ -627,6 +646,16 @@ def __lt__(self, other):
 class SInt(Int):
     hwtypes_T = ht.SIntVector
 
+    @debug_wire
+    def wire(self, other, debug_info=None):
+        if isinstance(other, UInt):
```
Comment (rsetaluri) [inline magma/bits.py:651]: it seems redundant to have logic `is_wireable` and then do this check again in `wire()`. Seems like `wire()` for any type should by default call `is_wireable()` and the error message of `"Cannot wire..."` is generic.

TL;DR: should we just call `is_wireable()` by default before any logic in `wire()`?

**Review by rsetaluri — COMMENTED** [review]
Thanks for addressing comments. Two last questions in comments.

File: magma/primitives/mux.py (hunk 3 above)
Comment (leonardt) [reply magma/primitives/mux.py:111]: Already removed by https://github.com/phanrahan/magma/pull/892/commits/69a2783b0110e230261c2d6d0af4005b9f9d762a

File: magma/bits.py (hunk 4 above)
Comment (leonardt) [reply magma/bits.py]: Ok, moved the wire error checking logic to wire() and updated to use is_wireable.  Overall I think it's an improvement, it consolidates the logic into a clear API, whereas before the error checking was done in various parts of the wiring pipeline.  As a side effect, it seems some error messages are improved, whereas other are less precise (since we early error before wiring things by checking is_wireable, we don't descend into the wiring pipeline where old errors were being raise).  I'll work on moving the old error checking and message logic up into the is_wireable flow (I think we can augment it to return an object that indicates and error with a message, this way we can report the generic error "Cannot wire X to Y" and then add a more specific error below it (e.g. array sizes do not match).

File: magma/bits.py [hunk 5]
```diff
@@ -23,6 +23,8 @@
 from magma.language_utils import primitive_to_python
 from magma.logging import root_logger
 from magma.generator import Generator2
+from magma.wire_container import WiringLog
+from magma.debug import debug_wire
```
Comment (rsetaluri) [inline magma/bits.py:27]: still need these 2? don't see them used

File: magma/wire.py [hunk 6]
```diff
@@ -52,5 +52,15 @@ def wire(o, i, debug_info=None):
...
 
+    i_T, o_T = type(i), type(o)
+    if not i_T.is_wireable(o_T):
+        _logger.error(
+            WiringLog(f"Cannot wire {{}} ({o_T}) to {{}} ({i_T})",
+                      o, i),
+            debug_info=debug_info
+        )
+        return
+
```
Comment (rsetaluri) [inline magma/wire.py:63]: nit: extra newline

**Review by rsetaluri — APPROVED** [review]
Looks good modulo final comments.

File: magma/bits.py (hunk 5 above)
Comment (leonardt) [reply magma/bits.py:27]: debug_wire is being used on line 159, don't need WiringLog anymore though
